buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -229,18 +229,111 @@ public static int hashCode(ByteBuf buffer) {
      * Returns the reader index of needle in haystack, or -1 if needle is not in haystack.
      */
     public static int indexOf(ByteBuf needle, ByteBuf haystack) {
-        // TODO: maybe use Boyer Moore for efficiency.
-        int attempts = haystack.readableBytes() - needle.readableBytes() + 1;
-        for (int i = 0; i < attempts; i++) {
-            if (equals(needle, needle.readerIndex(),
-                       haystack, haystack.readerIndex() + i,
-                       needle.readableBytes())) {
-                return haystack.readerIndex() + i;
+        if (haystack == null || needle == null) {
```

Comment:
Add a comment somewhere that says this is the Two Way algorithm, so those who are unfamiliar with it can learn more if they come across this code and get confused.

Reply:
and maybe add a link to the page that explains the algorithm nicely. There's one at wikipedia, but it seems a little bit dry - https://en.wikipedia.org/wiki/Two-way_string-matching_algorithm so please feel free to use a better one.